{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "de12a3f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "import json\n",
    "import os\n",
    "import csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a0d03356",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_DIR = \"/Users/user/Documents/Clades/July_2021\"\n",
    "label = \"C.37\"\n",
    "\n",
    "ANNOTATION = os.path.join(DATA_DIR, label+\"_annotation.json\")\n",
    "SUMMARY = os.path.join(DATA_DIR, label + \"_summary.json\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "4698368f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def map_to_gene(site):\n",
    "    global SUMMARY\n",
    "    with open(SUMMARY) as f:\n",
    "        data = json.load(f)\n",
    "    #end with\n",
    "    key = \"\"\n",
    "    for key in data:\n",
    "        #print(key)\n",
    "        theMap = data[key][\"map\"]\n",
    "        #rint(data[key][\"map\"])\n",
    "        #print(type(theMap))\n",
    "        #print(theMap)\n",
    "        if int(site) in theMap:\n",
    "            #print(\"Returning\", key)\n",
    "            return key, theMap.index(int(site))\n",
    "        #end if\n",
    "    return key, 0\n",
    "#end method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "89b48841",
   "metadata": {},
   "outputs": [],
   "source": [
    "NameMap = {\"S\": \"s2Spike\", \"E\": \"s2Envelope\", \"N\": \"s2Nucleocapsid\", \"ORF3a\": \"s2ORF3a\", \"methyltransferase\": \"s2Methyltransferase\",\n",
    "           \"helicase\": \"s2Helicase\", \"leader\": \"s2leader\", \"nsp10\": \"s2nsp10\",\n",
    "          \"nsp2\": \"s2nsp2\", \"nsp3\": \"s2nsp3\", \"nsp4\": \"s2nsp4\", \"nsp6\": \"s2nsp6\", \n",
    "           \"nsp7\": \"s2nsp7\", \"nsp8\": \"s2nsp8\", \"nsp9\": \"s2nsp9\",\n",
    "          \"ORF8\": \"s2ORF8\", \"ORF6\": \"s2ORF6\", \"3C\": \"s23C\", \"ORF7a\": \"s2ORF7a\",\n",
    "          \"exonuclease\": \"s2Exonuclease\", \"M\": \"s2Matrix\", \"endornase\":\"s2Endornase\", \"RdRp\": \"s2RdRP\"}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8a42e1a",
   "metadata": {},
   "source": [
    "## Version 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9cb18b3",
   "metadata": {},
   "source": [
    "## The whole table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "cd1469b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://observablehq.com/@aglucaci/visualizing-selection-analysis-results-for-evolution-of-t/8\n",
    "\n",
    "# Would be useful to have Table 1 results for spike\n",
    "# also Table 2 (cfel)\n",
    "# also the clade defining mutations\n",
    "\n",
    "# return:\n",
    "#.   genomic coordinate\n",
    "#.   Codon in gene\n",
    "#.   Num. selected branches? For meme not cfel\n",
    "#.   p-value for meme, cfel\n",
    "#.   q-value for meme, cfel sites\n",
    "#.   Domain for site, use ncbi annotation\n",
    "#.   AA frequency in clade\n",
    "#.   AA frequency in reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d1580424",
   "metadata": {},
   "outputs": [],
   "source": [
    "Gene = \"N\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "37c46c80",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output to csv\n",
    "def output_csv(data):\n",
    "    global label, Gene\n",
    "    with open(label + \"_\" + Gene +\".csv\", 'w') as csvfile:\n",
    "        #pamwriter = csv.writer(csvfile, delimiter=' ', quotechar='|', quoting=csv.QUOTE_MINIMAL)\n",
    "        spamwriter = csv.writer(csvfile)\n",
    "\n",
    "        for row in data:\n",
    "            spamwriter.writerow(row)\n",
    "            #spamwriter.writerow(['Spam'] * 5 + ['Baked Beans'])\n",
    "            #spamwriter.writerow(['Spam', 'Lovely Spam', 'Wonderful Spam'])\n",
    "        #end for\n",
    "    #end with\n",
    "#end method\n",
    "\n",
    "def AA_frequencies(AA_dict):\n",
    "    total = 0\n",
    "    results = {}\n",
    "    for item in AA_dict:\n",
    "        total += AA_dict.get(item, 0)\n",
    "    \n",
    "    for item in AA_dict:\n",
    "        AA = item\n",
    "        freq = AA_dict.get(item, 0) / total\n",
    "        results[AA] = round(freq,4)\n",
    "        \n",
    "    return results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "fb02cbb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Domain annotation\n",
    "# https://www.ncbi.nlm.nih.gov/nuccore/NC_045512\n",
    "# https://www.ncbi.nlm.nih.gov/protein/1796318598\n",
    "\n",
    "SPIKE_DOMAINS = {\"NTD\": [13, 304], \"RBD\": [319, 541], \"Furin\": [685, 656], \"HR1\": [918, 983], \"HR2\":[1162, 1203]}\n",
    "# there are additional domains to add in https://www.ncbi.nlm.nih.gov/protein/1796318598\n",
    "# also \"Sites\" like glycosylation sites.\n",
    "CD = {\"B-1-617-1\": {\"s2Spike\": [452, 484, 614, 681, 1071],\n",
    "                   \"s2ORF3a\": [26], \"s2ORF7a\":[82], \"s2Nucleocapsid\": [203, 377], \"s2nsp2\":[342, 397],\n",
    "                   \"s2RdRP\":[192, 538, 857]}, \n",
    "      \n",
    "      \"B-1-617-2\":{ \"s2RdRP\":[314, 662, 1000], \n",
    "                   \"s2Spike\": [19, 157,158, 452,478,614,681,950],\n",
    "                   \"s2ORF3a\": [26],\n",
    "                   \"s2Matrix\": [82],\n",
    "                  \"s2ORF7a\": [82, 120],\n",
    "                  \"s2ORF8\": [119, 120],\n",
    "                  \"s2Nucleocapsid\": [63, 203, 377],},\n",
    "      \"C.37\": {\"s2Spike\": [75, 76,452,490,614,859], \"s2Nucleocapsid\": [13, 203, 204, 214]}}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f225b52e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_sites(sl):\n",
    "    global output_data, data, Gene\n",
    "    #for site in sl: # site-list\n",
    "    print(\"# Processing:\", sl)\n",
    "    for key in data:\n",
    "        if data[key][\"G\"] != Gene:\n",
    "            continue\n",
    "        #end if\n",
    "        #print(key, int(data[key][\"S\"]))\n",
    "        #print(data[key][Gene])\n",
    "        if int(data[key][\"S\"]) in sl: #site-list, S here stands for Site, not Spike\n",
    "            #process it\n",
    "            #print(\"found site\")\n",
    "            pval = data[key].get(\"bMEME\", 1)\n",
    "            if pval != 1: pval = data[key][\"bMEME\"][\"p\"]\n",
    "            if pval <= 1.0:\n",
    "                x, i = map_to_gene(key)\n",
    "                branches_selected = str(data[key][\"bMEME\"][\"br\"])\n",
    "                qval = str(-1)\n",
    "                genomic_pos = key\n",
    "                reference_tag = \"Reference\"\n",
    "                AA_freq_clade = data[genomic_pos][\"aa\"][label]\n",
    "                AA_freq_reference = data[genomic_pos][\"aa\"][reference_tag]\n",
    "                AA_freq_clade = AA_frequencies(AA_freq_clade)\n",
    "                AA_freq_reference = AA_frequencies(AA_freq_reference)\n",
    "                h = AA_freq_clade\n",
    "                AA_freq_clade = \"\"\n",
    "                for item in h.keys():\n",
    "                    AA_freq_clade += item + \":\" + str(h[item]) + \" / \"\n",
    "                j = AA_freq_reference\n",
    "                AA_freq_reference = \"\"\n",
    "                for item in j.keys():\n",
    "                    AA_freq_reference += item + \":\" + str(j[item]) + \" / \"\n",
    "                # Domain\n",
    "                domain = \"S1\"\n",
    "                site = int(i) + 1\n",
    "                for region in SPIKE_DOMAINS.keys():\n",
    "                    start = SPIKE_DOMAINS[region][0]\n",
    "                    end = SPIKE_DOMAINS[region][1]\n",
    "                    if start <= site <= end:\n",
    "                        domain = region\n",
    "                        break\n",
    "                    #end if\n",
    "                if site >= 543: \n",
    "                    domain = \"S2\"  \n",
    "                output = [key, str(site), domain, branches_selected, str(round(pval, 4)), qval, AA_freq_clade, AA_freq_reference]\n",
    "                #print(\",\".join(output))\n",
    "                print(\"# adding\", output)\n",
    "                output_data.append(output)\n",
    "            #end if\n",
    "        #end if\n",
    "    #end for\n",
    "#end method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b8c38989",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GenomicPosition', 'Site', 'Domain', 'Num.BranchesSelected', 'p-value', 'q-value', 'AA Freq.Clade', 'AA Freq.Reference', 'CladeKeys', 'RefKeys']\n",
      "# Checking sites: [13, 203, 204, 214]\n",
      "# Processing: [13, 203, 204, 214]\n",
      "# adding ['28309', '13', 'NTD', '2', '0.0889', '-1', 'P:0.0588 / L:0.916 / F:0.0126 / I:0.0042 / -:0.0084 / ', 'P:0.952 / L:0.024 / S:0.008 / T:0.008 / H:0.008 / ']\n",
      "# adding ['28879', '203', 'NTD', '0', '0.6667', '-1', 'K:0.9832 / -:0.0084 / R:0.0084 / ', 'R:0.744 / K:0.24 / M:0.008 / G:0.008 / ']\n",
      "# adding ['28882', '204', 'NTD', '0', '0.6667', '-1', 'R:0.9832 / -:0.0084 / G:0.0084 / ', 'G:0.768 / T:0.008 / L:0.008 / R:0.2 / -:0.016 / ']\n",
      "# adding ['28912', '214', 'NTD', '0', '0.2152', '-1', 'G:0.0462 / S:0.0042 / C:0.937 / -:0.0126 / ', 'G:0.992 / C:0.008 / ']\n",
      "28303,11,S1,1,0.0458,-1,N:0.9832 / T:0.0042 / -:0.0042 / D:0.0084 / ,N:0.992 / S:0.008 / ,dict_keys(['N', 'T', '-', 'D']),dict_keys(['N', 'S'])\n",
      "28324,18,NTD,1,0.0378,-1,G:0.9874 / V:0.0084 / -:0.0042 / ,G:0.976 / V:0.024 / ,dict_keys(['G', 'V', '-']),dict_keys(['G', 'V'])\n",
      "28342,24,NTD,1,0.0309,-1,T:0.9832 / I:0.0168 / ,T:0.992 / I:0.008 / ,dict_keys(['T', 'I']),dict_keys(['T', 'I'])\n",
      "28351,27,NTD,1,0.0258,-1,N:0.9916 / D:0.0084 / ,N:1.0 / ,dict_keys(['N', 'D']),dict_keys(['N'])\n",
      "28417,49,NTD,1,0.0304,-1,T:0.9916 / N:0.0084 / ,T:1.0 / ,dict_keys(['T', 'N']),dict_keys(['T'])\n",
      "28540,90,NTD,1,0.0392,-1,A:0.9916 / S:0.0084 / ,A:0.992 / S:0.008 / ,dict_keys(['A', 'S']),dict_keys(['A', 'S'])\n",
      "28627,119,NTD,1,0.0189,-1,A:0.916 / P:0.0798 / S:0.0042 / ,A:0.968 / S:0.024 / V:0.008 / ,dict_keys(['A', 'P', 'S']),dict_keys(['A', 'S', 'V'])\n",
      "28675,135,NTD,1,0.0309,-1,T:0.9916 / I:0.0084 / ,T:0.976 / I:0.024 / ,dict_keys(['T', 'I']),dict_keys(['T', 'I'])\n",
      "28723,151,NTD,1,0.0432,-1,P:0.9748 / H:0.0042 / S:0.0126 / T:0.0042 / L:0.0042 / ,P:1.0 / ,dict_keys(['P', 'H', 'S', 'T', 'L']),dict_keys(['P'])\n",
      "28789,173,NTD,2,0.0323,-1,A:0.9748 / -:0.0042 / V:0.021 / ,A:1.0 / ,dict_keys(['A', '-', 'V']),dict_keys(['A'])\n",
      "28876,202,NTD,1,0.0003,-1,S:0.9874 / -:0.0126 / ,S:0.968 / N:0.024 / C:0.008 / ,dict_keys(['S', '-']),dict_keys(['S', 'N', 'C'])\n",
      "28906,212,NTD,1,0.0357,-1,G:0.9832 / V:0.0126 / -:0.0042 / ,G:0.952 / V:0.04 / S:0.008 / ,dict_keys(['G', 'V', '-']),dict_keys(['G', 'V', 'S'])\n",
      "29368,366,RBD,7,0.0005,-1,T:0.6261 / -:0.063 / I:0.3109 / ,T:0.984 / I:0.016 / ,dict_keys(['T', '-', 'I']),dict_keys(['T', 'I'])\n",
      "29371,367,RBD,2,0.0003,-1,E:0.8571 / -:0.1303 / T:0.0042 / G:0.0084 / ,E:0.992 / D:0.008 / ,dict_keys(['E', '-', 'T', 'G']),dict_keys(['E', 'D'])\n",
      "29413,381,RBD,1,0.0385,-1,A:0.8109 / -:0.1849 / S:0.0042 / ,A:0.992 / S:0.008 / ,dict_keys(['A', '-', 'S']),dict_keys(['A', 'S'])\n",
      "29464,398,RBD,2,0.0333,-1,A:0.7731 / -:0.2185 / V:0.0084 / ,A:0.984 / V:0.016 / ,dict_keys(['A', '-', 'V']),dict_keys(['A', 'V'])\n",
      "29473,401,RBD,0,0.0085,-1,D:0.7983 / -:0.1933 / A:0.0042 / Y:0.0042 / ,D:0.968 / Y:0.032 / ,dict_keys(['D', '-', 'A', 'Y']),dict_keys(['D', 'Y'])\n",
      "29479,403,RBD,0,0.0003,-1,F:0.7983 / -:0.1975 / H:0.0042 / ,F:1.0 / ,dict_keys(['F', '-', 'H']),dict_keys(['F'])\n",
      "29518,416,RBD,1,0.0001,-1,S:0.8529 / -:0.1387 / L:0.0042 / H:0.0042 / ,S:1.0 / ,dict_keys(['S', '-', 'L', 'H']),dict_keys(['S'])\n",
      "# Processing: []\n",
      "# writing\n",
      "# done.\n"
     ]
    }
   ],
   "source": [
    "# Get MEME Sites  -- Internal branches -- Table 1\n",
    "output_data = []\n",
    "\n",
    "# header\n",
    "header = [\"GenomicPosition\", \"Site\", \"Domain\", \"Num.BranchesSelected\", \"p-value\", \"q-value\", \"AA Freq.Clade\", \"AA Freq.Reference\", \"CladeKeys\", \"RefKeys\"]\n",
    "print(header)\n",
    "output_data.append(header)\n",
    "\n",
    "# initialize data\n",
    "with open(ANNOTATION) as f:\n",
    "    data = json.load(f)\n",
    "#end with\n",
    "\n",
    "# Do Clade defining sites\n",
    "if Gene == \"N\":\n",
    "    print(\"# Checking sites:\", CD[label][\"s2Nucleocapsid\"])\n",
    "    process_sites(CD[label][\"s2Nucleocapsid\"]) \n",
    "else:\n",
    "    print(\"# Checking sites:\", CD[label][\"s2Spike\"])\n",
    "    process_sites(CD[label][\"s2Spike\"])\n",
    "#end if\n",
    "\n",
    "#cd_sl = process_sites(CD[label][\"s2Spike\"])\n",
    "#for item in cd_sl: output_data.append(item)\n",
    "\n",
    "for key in data: # keys are the sites, genomic position\n",
    "    pval = 0\n",
    "    pval = data[key].get(\"bMEME\", 1)\n",
    "    if pval != 1: \n",
    "        pval = data[key][\"bMEME\"][\"p\"]\n",
    "    #end if\n",
    "    if pval <= 0.05:\n",
    "        x, i = map_to_gene(key)\n",
    "        if x != Gene: \n",
    "            continue # only want spike or nucleocapsid\n",
    "        #end if\n",
    "        branches_selected = str(data[key][\"bMEME\"][\"br\"])\n",
    "        qval = str(-1)\n",
    "        #domain = \"N/A\"\n",
    "        \n",
    "        genomic_pos = key\n",
    "        reference_tag = \"Reference\"\n",
    "        AA_freq_clade = data[genomic_pos][\"aa\"][label]\n",
    "        AA_freq_reference = data[genomic_pos][\"aa\"][reference_tag]\n",
    "        AA_freq_clade = AA_frequencies(AA_freq_clade)\n",
    "        AA_freq_reference = AA_frequencies(AA_freq_reference)\n",
    "        h = AA_freq_clade\n",
    "        AA_freq_clade = \"\"\n",
    "        for item in h.keys():\n",
    "            AA_freq_clade += item + \":\" + str(h[item]) + \" / \"\n",
    "            \n",
    "        j = AA_freq_reference\n",
    "        AA_freq_reference = \"\"\n",
    "        for item in j.keys():\n",
    "            AA_freq_reference += item + \":\" + str(j[item]) + \" / \"\n",
    "            \n",
    "        # Domain\n",
    "        domain = \"S1\"\n",
    "        site = int(i)+1\n",
    "        \n",
    "        for region in SPIKE_DOMAINS.keys():\n",
    "            start = SPIKE_DOMAINS[region][0]\n",
    "            end = SPIKE_DOMAINS[region][1]\n",
    "            if start <= site <= end:\n",
    "                domain = region\n",
    "                break\n",
    "        \n",
    "        if site >= 543: domain = \"S2\"  \n",
    "        output = [key, str(site), domain, branches_selected, str(round(pval, 4)), \n",
    "                  qval, AA_freq_clade, AA_freq_reference, \n",
    "                  str(h.keys()), str(j.keys())]\n",
    "        print(\",\".join(output))\n",
    "        output_data.append(output)\n",
    "    #end if\n",
    "#end for\n",
    "        \n",
    "# Do cfel here.\n",
    "cfel_sites = []\n",
    "\n",
    "for key in data:\n",
    "    pval = data[key][\"bCFEL\"][\"pp\"]\n",
    "    if abs(pval) <= 0.1:\n",
    "        x, i = map_to_gene(key)\n",
    "        if x != \"S\": continue\n",
    "        #print(x, NameMap[x], int(i)+1, int(i)+1, 0.599)\n",
    "        cfel_sites.append (int(i)+1)\n",
    "process_sites(cfel_sites)\n",
    "\n",
    "print(\"# writing\")     \n",
    "output_csv(output_data)\n",
    "print(\"# done.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "63e95321",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|    |   GenomicPosition |   Site | Domain   |   Num.BranchesSelected |   p-value | AA Freq.Clade                                          | AA Freq.Reference                                 |\n",
      "|---:|------------------:|-------:|:---------|-----------------------:|----------:|:-------------------------------------------------------|:--------------------------------------------------|\n",
      "|  1 |             28303 |     11 | S1       |                      1 |    0.0458 | N:0.9832 / T:0.0042 / -:0.0042 / D:0.0084 /            | N:0.992 / S:0.008 /                               |\n",
      "|  2 |             28309 |     13 | NTD      |                      2 |    0.0889 | P:0.0588 / L:0.916 / F:0.0126 / I:0.0042 / -:0.0084 /  | P:0.952 / L:0.024 / S:0.008 / T:0.008 / H:0.008 / |\n",
      "|  3 |             28324 |     18 | NTD      |                      1 |    0.0378 | G:0.9874 / V:0.0084 / -:0.0042 /                       | G:0.976 / V:0.024 /                               |\n",
      "|  4 |             28342 |     24 | NTD      |                      1 |    0.0309 | T:0.9832 / I:0.0168 /                                  | T:0.992 / I:0.008 /                               |\n",
      "|  5 |             28351 |     27 | NTD      |                      1 |    0.0258 | N:0.9916 / D:0.0084 /                                  | N:1.0 /                                           |\n",
      "|  6 |             28417 |     49 | NTD      |                      1 |    0.0304 | T:0.9916 / N:0.0084 /                                  | T:1.0 /                                           |\n",
      "|  7 |             28540 |     90 | NTD      |                      1 |    0.0392 | A:0.9916 / S:0.0084 /                                  | A:0.992 / S:0.008 /                               |\n",
      "|  8 |             28627 |    119 | NTD      |                      1 |    0.0189 | A:0.916 / P:0.0798 / S:0.0042 /                        | A:0.968 / S:0.024 / V:0.008 /                     |\n",
      "|  9 |             28675 |    135 | NTD      |                      1 |    0.0309 | T:0.9916 / I:0.0084 /                                  | T:0.976 / I:0.024 /                               |\n",
      "| 10 |             28723 |    151 | NTD      |                      1 |    0.0432 | P:0.9748 / H:0.0042 / S:0.0126 / T:0.0042 / L:0.0042 / | P:1.0 /                                           |\n",
      "| 11 |             28789 |    173 | NTD      |                      2 |    0.0323 | A:0.9748 / -:0.0042 / V:0.021 /                        | A:1.0 /                                           |\n",
      "| 12 |             28876 |    202 | NTD      |                      1 |    0.0003 | S:0.9874 / -:0.0126 /                                  | S:0.968 / N:0.024 / C:0.008 /                     |\n",
      "| 13 |             28879 |    203 | NTD      |                      0 |    0.6667 | K:0.9832 / -:0.0084 / R:0.0084 /                       | R:0.744 / K:0.24 / M:0.008 / G:0.008 /            |\n",
      "| 14 |             28882 |    204 | NTD      |                      0 |    0.6667 | R:0.9832 / -:0.0084 / G:0.0084 /                       | G:0.768 / T:0.008 / L:0.008 / R:0.2 / -:0.016 /   |\n",
      "| 15 |             28906 |    212 | NTD      |                      1 |    0.0357 | G:0.9832 / V:0.0126 / -:0.0042 /                       | G:0.952 / V:0.04 / S:0.008 /                      |\n",
      "| 16 |             28912 |    214 | NTD      |                      0 |    0.2152 | G:0.0462 / S:0.0042 / C:0.937 / -:0.0126 /             | G:0.992 / C:0.008 /                               |\n",
      "| 17 |             29368 |    366 | RBD      |                      7 |    0.0005 | T:0.6261 / -:0.063 / I:0.3109 /                        | T:0.984 / I:0.016 /                               |\n",
      "| 18 |             29371 |    367 | RBD      |                      2 |    0.0003 | E:0.8571 / -:0.1303 / T:0.0042 / G:0.0084 /            | E:0.992 / D:0.008 /                               |\n",
      "| 19 |             29413 |    381 | RBD      |                      1 |    0.0385 | A:0.8109 / -:0.1849 / S:0.0042 /                       | A:0.992 / S:0.008 /                               |\n",
      "| 20 |             29464 |    398 | RBD      |                      2 |    0.0333 | A:0.7731 / -:0.2185 / V:0.0084 /                       | A:0.984 / V:0.016 /                               |\n",
      "| 21 |             29473 |    401 | RBD      |                      0 |    0.0085 | D:0.7983 / -:0.1933 / A:0.0042 / Y:0.0042 /            | D:0.968 / Y:0.032 /                               |\n",
      "| 22 |             29479 |    403 | RBD      |                      0 |    0.0003 | F:0.7983 / -:0.1975 / H:0.0042 /                       | F:1.0 /                                           |\n",
      "| 23 |             29518 |    416 | RBD      |                      1 |    0.0001 | S:0.8529 / -:0.1387 / L:0.0042 / H:0.0042 /            | S:1.0 /                                           |\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv(label + \"_\" + Gene +\".csv\")\n",
    "#df.index += 1\n",
    "df = df.drop(columns=['q-value', 'CladeKeys', 'RefKeys'])\n",
    "sorted_df = df.sort_values(by='Site', ascending=True)\n",
    "#df = df.rename_axis('MyIdx').sort_values(by = ['MyCol', 'MyIdx'], ascending = [False, True])\n",
    "sorted_df = sorted_df.reset_index(drop=True)\n",
    "sorted_df.index += 1\n",
    "print(sorted_df.to_markdown())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d0f38c2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
